### ISSUE TITLE
`define-macros-order`: should allow referencing local variables defined before the macros statement

### ISSUE THREAD

**Issue by andreww2012** [vuejs/eslint-plugin-vue#2746]
**What did you do?**

`define-macros-order` should probably allow referencing local variables defined before the macros:

```vue
<script setup lang="ts">
/* eslint vue/define-macros-order: [2, {order: ["defineOptions", "defineModel", "defineProps", "defineEmits", "defineSlots"]}] no-use-before-define: 2 */
import {ref} from 'vue';

const lastSetModelValue = ref<string>();
// "defineModel should be the first statement in `<script setup>` (after any potential import statements or type definitions)",
// but it's not safe to move `lastSetModelValue` after the macros
const [modelValue, modelModifiers] = defineModel<string>({
  set: (value: string) => {
    lastSetModelValue.value = value;
    return value;
  },
  get: (value: string) => {
    return lastSetModelValue.value ?? value;
  },
});
</script>
```

**What did you expect to happen?**

The rule should not report .

If `lastSetModelValue` definition is moved below `defineModel`, [`no-use-before-define`](https://eslint.org/docs/latest/rules/no-use-before-define) will be triggered as well as `define-macros-order` on the following macros definitions.

**Repository to reproduce this issue**

[Reproduction](https://ota-meshi.github.io/eslint-plugin-vue-demo/#eJyNU02PmzAQ/SsjX/ZDuEg5skm2lx6rVlqplyUHQobEqrHd8ZCqQvz3jiFhyVatesHWe2++PI9e1f6AqlDrWJMJDBG5C2Ard9yUimOptqXLHwGjNY7h3GF+wMY41G1Vk4/a0wGpgNdVBv31XqpJ8yWw8U5yZHCFPks1uwS+kg83ik+t4RvgxfoE7IYdOK+7iHqPjSfUE13ACh7z0pk2eGLoCZsBGvIt3Em3d0+lkwFyqKURlrkivyCPXXyrbIewAQlYRybjjtv7B5FPytd21mQw3iXINAYp7iRoMc0c3JcO0vsVcH9OcQVMzANstjCS8GcDH86XNsbzaVKRLIHcAhqy9D3+O/cl6m8lnp/fJRzepv2fd7lqQ9rY/ATj/m54TAuc+XGdI7/OJ4uJoVSmqLMYVdGrZKmKpc6+Y7z4SRWrbCT21tff32FiAuvFgIxtsBXjkmg7yyZY1FFMoys6xiVL+KMzhAedRtA/DZ+ShyqJWarO+sStnZGxuDZO114c5tDxW04+mZioRStDpkJFMfXbqyRVH/lXwGl0Pf1G+UUxDL8BR6g4UQ==)

**Comment by ota-meshi** [comment]
> but it's not safe to move `lastSetModelValue` after the macros

Why is it not safe? Could you provide a concrete example where you can see that it is not safe?

**Comment by andreww2012** [comment]
I think my wording wasn't fully correct - it only *might* be unsafe because you're required to ensure any potential accesses to `modelValue` performed before the initialization of `lastSetModelValue`, which makes your code more fragile, and you also get reports from `no-use-before-define` as well as lose the ability to put the related code together.

[Vue playground example](https://play.vuejs.org/#eNqNU01v2zAM/SuELmmBwj5spyxJsQ09dMA+sA4DhmlANJt21MmSIdFJisD/fZQcJ0E/gl5siXzv8ZG0d+J922brDsVUzELhdUsQkLoWjLL1XAoKUiyk1U3rPMFH17RQedfAJMvjJVIn76Sd5QOZoXwhbFqjCPkGMEuknM+z/CQhrli7cLbSdXYfnGUDuwiXomC8Nui/tqSd5fpTSJmYU8a4zacUI9/h1RgvVlj8eyZ+H7YxJsU3jwH9GqU45Ej5GmlI39x9wS2fD8nGlZ1h9JnkdwzOdNHjAPvQ2ZJtn+CS29s0Om3rH+FmS2jD2FQ0GpF9wkvBo4yjeqn1o9032dvEk7bnKY5reN0Gdx6rfr/C/e6k5TUEgt/cFpqfyvD8IJ0/u1JXGn34A3MosdIWOYRmFshzQ4uLZI6rTeFiHXlTGDKXMF+Mzo0KdIeUiEk8S1BWTG82EFGeLXt7Ehqmwh2f096zXipxff1IsL9M/eY53Daq5naAVor4gR5BB1BgHIGroGApSFFly5i38OA6sIh8c6BC0PUgpDi4GcrEzPI4xGUGv5iTyCPRY4PNX/RJxGzUQ0giJRZGcbHlk06WoCqK+FW0l8o2aIkbkvZY6jDSyelCn4gxgNd/2F4cxpkfN35Yj3/a/j9r0276)

**Comment by ota-meshi** [comment]
So it's technically safe, right? If so, I think it's a stylistic issue, and it would be good to add an option to handle it.
Please open a PR if you're interested.
